ext/phar/phar.c

```diff
@@ -600,49 +570,161 @@ int phar_open_parsed_phar(char *fname, size_t fname_len, char *alias, size_t ali
+			smart_str main_metadata_str = {0};
+			php_serialize_data_t metadata_hash;
+			ZEND_ASSERT(!persistent);
+			PHP_VAR_SERIALIZE_INIT(metadata_hash);
+			php_var_serialize(&main_metadata_str, &tracker->val, &metadata_hash);
+			PHP_VAR_SERIALIZE_DESTROY(metadata_hash);
+			if (EG(exception)) {
+				return FAILURE;
+			}
+			tracker->str = main_metadata_str.s;
+		}
+		/* precondition: This has data */
+		ZEND_ASSERT(tracker->str != NULL);
 		ZVAL_NULL(metadata);
-		PHP_VAR_UNSERIALIZE_INIT(var_hash);
+		start = ZSTR_VAL(tracker->str);
 
-		if (!php_var_unserialize(metadata, &p, p + zip_metadata_len, &var_hash)) {
-			efree(p_buff);
-			PHP_VAR_UNSERIALIZE_DESTROY(var_hash);
+		unserialize_with_options(metadata, start, ZSTR_LEN(tracker->str), unserialize_options, method_name);
+		if (EG(exception)) {
 			zval_ptr_dtor(metadata);
 			ZVAL_UNDEF(metadata);
 			return FAILURE;
 		}
-		efree(p_buff);
-		PHP_VAR_UNSERIALIZE_DESTROY(var_hash);
-
-		if (PHAR_G(persist)) {
-			/* lazy init metadata */
-			zval_ptr_dtor(metadata);
-			Z_PTR_P(metadata) = pemalloc(zip_metadata_len, 1);
-			memcpy(Z_PTR_P(metadata), *buffer, zip_metadata_len);
-			return SUCCESS;
-		}
+		return SUCCESS;
 	} else {
-		ZVAL_UNDEF(metadata);
+		/* TODO: what is the current/expected behavior when fetching an object set with setMetadata then getting it
+		 * with getMetadata() and modifying a property? Previously, it was underdefined, and probably unimportant to support. */
+		ZVAL_COPY(metadata, &tracker->val);
 	}
 
 	return SUCCESS;
 }
 /* }}}*/
 
+/**
+ * Check if this has any data, serialized or as a raw value.
+ */
+zend_bool phar_metadata_tracker_has_data(const phar_metadata_tracker *tracker, int persistent) /* {{{ */
+{
+	ZEND_ASSERT(!persistent || Z_ISUNDEF(tracker->val));
+	return !Z_ISUNDEF(tracker->val) || tracker->str != NULL;
+}
+/* }}} */
+
+/**
+ * Free memory used to track the metadata and set all fields to be null/undef.
+ */
+void phar_metadata_tracker_free(phar_metadata_tracker *tracker, int persistent) /* {{{ */
+{
+	if (!Z_ISUNDEF(tracker->val)) {
+		ZEND_ASSERT(!persistent);
+		zval_ptr_dtor(&tracker->val);
+		ZVAL_UNDEF(&tracker->val);
+	}
+	if (tracker->str) {
+		zend_string_release(tracker->str);
+		tracker->str = NULL;
+	}
+}
+/* }}} */
+
+/**
+ * Free memory used to track the metadata and set all fields to be null/undef.
+ */
+void phar_metadata_tracker_copy(phar_metadata_tracker *dest, const phar_metadata_tracker *source, int persistent) /* {{{ */
+{
+	ZEND_ASSERT(dest != source);
+	phar_metadata_tracker_free(dest, persistent);
+
+	if (!Z_ISUNDEF(source->val)) {
+		ZEND_ASSERT(!persistent);
+		ZVAL_COPY(&dest->val, &source->val);
+	}
+	if (source->str) {
+		dest->str = zend_string_copy(source->str);
+	}
+}
+/* }}} */
+
+/**
+ * Increment reference counts after a metadata entry was copied
+ */
+void phar_metadata_tracker_clone(phar_metadata_tracker *tracker) /* {{{ */
+{
+	zval_copy_ctor(&tracker->val);
```

Comment:
```suggestion
	Z_TRY_ADDREF_P(&tracker->val);
```

Unless this was intended to perform a duplication?

Reply:
This was intended to be equivalent to the original code, e.g. in the way (top-level?) arrays are duplicated.

I can't tell why it was done that way - that may have just been an inefficiency in the original code